src/scoperesolver.js

```diff
@@ -236,192 +243,60 @@ class ScopeResolver extends events.EventEmitter {
-      // already satisfied by the scopeset.  This is equivalent to `seen =
-      // mergeScopeSets(seen, [scope])`, then returning false if seen is not
-      // changed.
-      let see = (scope) => {
-        const n = seen.length;
-        const trailingStar = scope.endsWith('*');
-        const prefix = scope.slice(0, -1);
-        let i = 0;
-        while (i < n) {
-          let seenScope = seen[i];
-          // if seenScope satisfies scope, we're done
-          if (scope === seenScope || seenScope.endsWith('*') && scope.startsWith(seenScope.slice(0, -1))) {
-            return false;
-          }
-
-          // if we've found where to insert this scope, do so and splice out any existing scopes
-          // that this one satisfies
-          if (scopeCompare(seenScope, scope) > 0) {
-            let j = i;
-            if (trailingStar) {
-              while (j < n && seen[j].startsWith(prefix)) {
-                j++;
-              }
-            }
-            seen.splice(i, j - i, scope);
-            return true;
-          }
-
-          i++;
-        }
-
-        // we fell off the end of `seen`, so add this new scope at the end
-        seen.push(scope);
-        return true;
-      };
-
-      let i = 0;
-      while (i < queue.length) {
-        let scope = queue[i++];
-
-        // execute the DFA and expand any parameterizations in the result, then add
-        // the newly expanded scopes to the list of scopes to expand (recursively)
-        const trailingStar = scope.endsWith('*');
-        executeTrie(dfa, scope, (scopes, offset) => {
-          // Get the replacement slice for any parameterization. If this is empty and the
-          // scope ended with `*`, consider that `*` to have been extended into the
-          // replacement.
-          const slice = scope.slice(offset) || (trailingStar ? '*' : '');
-          const parameter_regexp = trailingStar ? PARAMETER_TO_END : PARAMETER_G;
-
-          scopes.map(s => s.replace(parameter_regexp, slice)).forEach(s => {
-            // mark this scope as seen, and if it is novel and expandable, add it
-            // to the queue for expansion
-            if (see(s) && ASSUME_PREFIX.test(s)) {
-              queue.push(s);
-            }
-          });
-        });
+      // Build expansion of queue
+      const result = new ScopeSetBuilder();
+      for (const scope of queue) {
+        const input = scope.startsWith('assume:') ? scope.slice(7) : '*';
+        trie.execute(node, input, result);
       }
+      // Store result in cache
+      const scopes = result.scopes();
+      lru.set(cacheKey, scopes);
 
-      lru.set(cacheKey, seen);
-      return mergeScopeSets(inputs, seen);
+      return ScopeSetBuilder.mergeScopeSets(scopes, inputs);
     };
-  };
+  }
 
   /**
    * Return a normalized set of scopes that `scopes` can be expanded to when
    * assuming all authorized roles.
    */
```

Comment:
It looks like the `static normalizeScopes` method is no longer used?

Reply:
I'm not sure it was used before... there is a few things like that. I prefer to cleanup later.